OS X compile fails with 'Invalid function: cdr'

When building on MacOS, assuming PR #37 is applied, the build terminates as follows
```
4480 unused bytes follow Mach-O header
89147 pure bytes used
mv -f remacs bootstrap-emacs
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../lisp compile-first EMACS="../src/bootstrap-emacs"
  ELC      emacs-lisp/macroexp.elc
Invalid function: eq
make[2]: *** [emacs-lisp/macroexp.elc] Error 255
make[1]: *** [bootstrap-emacs] Error 2
make: *** [src] Error 2
```

`eq` appears to be defined in `rust_src/src/strings.rs` but doesn't make it in to `bootstrap-emacs` as a symbol
```
$ strings src/bootstrap-emacs | grep eq
[ not there ]
```
whereas `stringp` which is also defined in `rust_src/src/strings.rs` is a symbol
```
$ strings src/bootstrap-emacs| grep stringp
stringp
stringp
```

This is on macOS, but from a clean start I now get `Invalid function: car`:
```
[ *snip* ]
mv -f remacs bootstrap-emacs
/Library/Developer/CommandLineTools/usr/bin/make -C ../lisp compile-first EMACS="../src/bootstrap-emacs"
  ELC      emacs-lisp/macroexp.elc
Invalid function: car
make[2]: *** [emacs-lisp/macroexp.elc] Error 255
make[1]: *** [bootstrap-emacs] Error 2
make: *** [src] Error 2
```
How do I go about debugging that? I can see likely symbols in `libremacs.a`
```
$ nm rust_src/target/release/libremacs.a | grep car
0000000000000810 T _Fcar
0000000000000770 T _Fsetcar
0000000000003940 b __ZN55_$LT$remacs..cons..Scar$u20$as$u20$core..ops..Deref$GT$5deref11__stability4LAZY17h359f9d3f18e67075E
0000000000003920 b __ZN58_$LT$remacs..cons..Ssetcar$u20$as$u20$core..ops..Deref$GT$5deref11__stability4LAZY17h8c55925ab1e59632E
```
and also in the executable
```
$ nm src/bootstrap-emacs | grep car
00000001001d7700 T _Fcar
00000001000f0b80 T _Fcar_less_than_car
000000010011d090 T _Fcar_safe
00000001000ba170 T _Fdiscard_input
000000010013cf80 T _Fmapcar
00000001001d7660 T _Fsetcar
000000010022f390 S _Qcar_less_than_car
00000001002b1068 d _Scar_less_than_car
00000001005e4118 d _Scar_safe
00000001002aee78 d _Sdiscard_input
00000001005e75a0 d _Smapcar
00000001006757b8 b __ZN55_$LT$remacs..cons..Scar$u20$as$u20$core..ops..Deref$GT$5deref11__stability4LAZY17h359f9d3f18e67075E
0000000100675798 b __ZN58_$LT$remacs..cons..Ssetcar$u20$as$u20$core..ops..Deref$GT$5deref11__stability4LAZY17h8c55925ab1e59632E
00000001001c83d0 T _careadlinkat
00000001000580b0 T _discard_menu_items
00000001000c2bb0 T _discard_mouse_events
00000001000d30b0 T _discard_tty_input
0000000100010b60 T _frames_discard_buffer
000000010013cce0 t _mapcar1
0000000100119660 t _mark_discard_killed_buffers
00000001000b3cf0 T _xcar_addr
$
```
and if I try to just run the exe
```
$ ./src/bootstrap-emacs
Invalid function: car
$
```
where do I go from here?

```
$ git diff
diff --git a/rust_src/src/lib.rs b/rust_src/src/lib.rs
index a51a0a8..b0e231a 100644
--- a/rust_src/src/lib.rs
+++ b/rust_src/src/lib.rs
@@ -39,6 +39,7 @@ extern "C" {
 #[no_mangle]
 pub extern "C" fn rust_init_syms() {
     unsafe {
+        println!("calling rust_init_syms");
         defsubr(&*atom::Satom);
         defsubr(&*math::Smod);
         defsubr(&*math::Splus);
$ make clean && make
[*snip*]
mv -f remacs bootstrap-emacs
/Library/Developer/CommandLineTools/usr/bin/make -C ../lisp compile-first EMACS="../src/bootstrap-emacs"
  ELC      emacs-lisp/macroexp.elc
Invalid function: car
make[2]: *** [emacs-lisp/macroexp.elc] Error 255
make[1]: *** [bootstrap-emacs] Error 2
make: *** [src] Error 2
$
```

Same thing as you I think.

```
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../admin/unidata all EMACS="../../src/bootstrap-emacs"
  ELC      uvs.elc
Invalid function: car
make[2]: *** [uvs.elc] Error 255
make[1]: *** [macuvs.h] Error 2
make: *** [src] Error 2
```

--- a/rust_src/Cargo.toml
+++ b/rust_src/Cargo.toml
@@ -11,6 +11,9 @@ build = "build.rs"
 [dependencies]
 lazy_static = "0.2.2"
 libc = "0.2.17"
+
+# Only want this local crate as dependency on Mac OS X
+[target.'cfg(target_os = "macos")'.dependencies]
 alloc_unexecmacosx = { version = "0.1.0", path = "alloc_unexecmacosx" }
 
 [build-dependencies]

--- a/rust_src/Cargo.toml
+++ b/rust_src/Cargo.toml
@@ -11,6 +11,7 @@ build = "build.rs"
 [dependencies]
 lazy_static = "0.2.2"
 libc = "0.2.17"
+alloc_unexecmacosx = { version = "0.1.0", path = "alloc_unexecmacosx" }
 
 [build-dependencies]
 libc = "0.2.17"

--- a/rust_src/alloc_unexecmacosx/Cargo.toml
+++ b/rust_src/alloc_unexecmacosx/Cargo.toml
@@ -1,0 +1,11 @@
+[package]
+name = "alloc_unexecmacosx"
+version = "0.1.0"
+authors = ["Felix S. Klock II <pnkfelix@pnkfx.org>"]
+
+# Would be nice to do this, but saying "default-features = false"
+# locally gets cancelled out by the other imports of libc globally.
+# Instead we will just cut-and-paste type definitions (boo!).
+#
+# [dependencies]
+# libc = { version = "0.2.17", default-features = false }

--- a/rust_src/alloc_unexecmacosx/src/lib.rs
+++ b/rust_src/alloc_unexecmacosx/src/lib.rs
@@ -45,24 +45,33 @@ extern "C" {
 // what is found below.
 //
 // Note that the standard `malloc` and `realloc` functions do not provide a way
-// to communicate alignment so this implementation would need to be improved
-// with respect to alignment in that aspect.
+// to communicate alignment. While a more ambitious implementor might have
+// attempted to support custom alignments, this simple code simply asserts that
+// we never request an alignment that is not already satisfied by the underlying
+// malloc call.
 
 #[no_mangle]
-pub extern fn __rust_allocate(size: usize, _align: usize) -> *mut u8 {
-    unsafe { unexec_malloc(size as libc::size_t) as *mut u8 }
+pub extern fn __rust_allocate(size: usize, align: usize) -> *mut u8 {
+    unsafe {
+        let addr = unexec_malloc(size as libc::size_t) as usize;
+        assert_eq!(addr & (align - 1), 0);
+        addr as *mut u8
+    }
 }
 
 #[no_mangle]
-pub extern fn __rust_deallocate(ptr: *mut u8, _old_size: usize, _align: usize) {
+pub extern fn __rust_deallocate(ptr: *mut u8, _old_size: usize, align: usize) {
+    assert_eq!(ptr as usize & (align - 1), 0);
     unsafe { unexec_free(ptr as *mut libc::c_void) }
 }
 
 #[no_mangle]
 pub extern fn __rust_reallocate(ptr: *mut u8, _old_size: usize, size: usize,
-                                _align: usize) -> *mut u8 {
+                                align: usize) -> *mut u8 {
     unsafe {
-        unexec_realloc(ptr as *mut libc::c_void, size as libc::size_t) as *mut u8
+        let addr = unexec_realloc(ptr as *mut libc::c_void, size as libc::size_t) as usize;
+        assert_eq!(addr & (align - 1), 0);
+        addr as *mut u8
     }
 }
 

--- a/rust_src/alloc_unexecmacosx/src/lib.rs
+++ b/rust_src/alloc_unexecmacosx/src/lib.rs
@@ -1,0 +1,78 @@
+#![feature(allocator)]
+#![allocator]
+
+#![no_std]
+
+// // Would be nice to do this, but saying "default-features = false"
+// // locally gets cancelled out by the other imports of libc globally.
+// // Instead we will just cut-and-paste type definitions (boo!).
+// //
+// // See also https://github.com/rust-lang/rust/issues/39262
+//
+// extern crate libc;
+
+mod libc {
+    #![allow(non_camel_case_types)]
+
+    pub type size_t = usize;
+
+    // Use repr(u8) as LLVM expects `void*` to be the same as `i8*` to help enable
+    // more optimization opportunities around it recognizing things like
+    // malloc/free.
+    #[repr(u8)]
+    pub enum c_void {
+        // Two dummy variants so the #[repr] attribute can be used.
+        #[doc(hidden)]
+        __variant1,
+        #[doc(hidden)]
+        __variant2,
+    }
+}
+
+/// To adhere to the rule that all calls to malloc, realloc, and free
+/// be redirected to their "unexec_"-prefixed variants, this crate
+/// provides a custom system allocator that performs such a mapping.
+
+extern "C" {
+    fn unexec_malloc(size: libc::size_t) -> *mut libc::c_void;
+    fn unexec_realloc(old_ptr: *mut libc::c_void, new_size: libc::size_t) -> *mut libc::c_void;
+    fn unexec_free(ptr: *mut libc::c_void);
+}
+
+// Listed below are the five allocation functions currently required by custom
+// allocators. Their signatures and symbol names are not currently typechecked
+// by the compiler, but this is a future extension and are required to match
+// what is found below.
+//
+// Note that the standard `malloc` and `realloc` functions do not provide a way
+// to communicate alignment so this implementation would need to be improved
+// with respect to alignment in that aspect.
+
+#[no_mangle]
+pub extern fn __rust_allocate(size: usize, _align: usize) -> *mut u8 {
+    unsafe { unexec_malloc(size as libc::size_t) as *mut u8 }
+}
+
+#[no_mangle]
+pub extern fn __rust_deallocate(ptr: *mut u8, _old_size: usize, _align: usize) {
+    unsafe { unexec_free(ptr as *mut libc::c_void) }
+}
+
+#[no_mangle]
+pub extern fn __rust_reallocate(ptr: *mut u8, _old_size: usize, size: usize,
+                                _align: usize) -> *mut u8 {
+    unsafe {
+        unexec_realloc(ptr as *mut libc::c_void, size as libc::size_t) as *mut u8
+    }
+}
+
+#[no_mangle]
+pub extern fn __rust_reallocate_inplace(_ptr: *mut u8, old_size: usize,
+                                        _size: usize, _align: usize) -> usize {
+    old_size // this api is not supported by libc
+}
+
+#[no_mangle]
+pub extern fn __rust_usable_size(size: usize, _align: usize) -> usize {
+    size
+}

--- a/rust_src/src/lib.rs
+++ b/rust_src/src/lib.rs
@@ -3,6 +3,8 @@
 
 #![cfg_attr(feature = "strict", deny(warnings))]
 
+// Wilfred/remacs#38 : Need to override the allocator for legacy unexec support on Mac.
+#[cfg(target_os = "macos")]
 extern crate alloc_unexecmacosx;
 
 #[macro_use]

--- a/rust_src/src/lib.rs
+++ b/rust_src/src/lib.rs
@@ -3,6 +3,8 @@
 
 #![cfg_attr(feature = "strict", deny(warnings))]
 
+extern crate alloc_unexecmacosx;
+
 #[macro_use]
 extern crate lazy_static;
 

